File: solutions/go/luhn/1/luhn.go
/*
Package luhn implements a simple Luhn algorithm check
*/
package luhn

import (
	"strconv"
	"strings"
)

//Valid returns true if it passes the Luhn algorithm
func Valid(number string) bool {
	number = strings.Replace(number, " ", "", -1)
	len := len(number)
	if len <= 1 {
		return false
	}
	parity := len % 2
	sum := int64(0)
	for i := 0; i < len; i++ {
		c := string([]rune(number)[i])
		if digit, err := strconv.ParseInt(c, 10, 32); err == nil {
			if i%2 == parity {
				digit *= 2
			}
			if digit > 9 {
				digit -= 9
			}
			sum += digit
		} else {
			// invalid digit char, not valid
			return false
		}
	}
	return sum%10 == 0
}

File: solutions/csharp/luhn/1/Luhn.cs
using System.Linq;
using System.Text.RegularExpressions;

public static class Luhn
{
    private static readonly Regex _regex = new(@"^(\d *){2,}$");
    
    public static bool IsValid(string number) => _regex.IsMatch(number) 
        && number
        .Replace(" ", "")
        .Reverse()
        .Select(Checksum)
        .Sum() % 10 == 0;
    
    public static int Checksum(char digit, int index) => index % 2 == 0 
        ? digit - '0' : (digit - '0') * 2 is var d && d > 9 
            ? d - 9 : d;
}